File: python/3085. Minimum Deletions to Make String K-Special.py
class Solution:
    def minimumDeletions(self, word: str, k: int) -> int:
        freq = [0] * 26
        for i in range(0, len(word)):
            freq[ord(word[i])-ord('a')] += 1
        f = []
        for i in freq:
            if i != 0:
                f.append(i)
        f.sort()
        return self.find(f, 0, len(f)-1, k)
    def find(self, f, left, right, k):
        if left >= right:
            return 0
        if f[right] - f[left] <= k:
            return 0
        return min(f[left] + self.find(f, left+1, right, k), f[right] - f[left] - k + self.find(f, left, right-1, k))

File: java/3085. Minimum Deletions to Make String K-Special.java
class Solution {
    public int minimumDeletions(String word, int k) {
        int[] freq = new int[26];
        for (int i = 0; i < word.length(); i++) {
            freq[word.charAt(i) - 'a']++;
        }
        List<Integer> f = new ArrayList<>();
        for (int i : freq) if (i != 0) f.add(i);
        Collections.sort(f);
        return find(f, 0, f.size()-1, k);
    }
    private int find(List<Integer> f, int left, int right, int k) {
        if (left >= right) return 0;
        if (f.get(right) - f.get(left) <= k) return 0;
        return Math.min(
            f.get(left) + find(f, left+1, right, k),
            f.get(right) - f.get(left) - k + find(f, left, right-1, k)
        );
    }
}
